core/src/cluster_nodes.rs

```diff
@@ -316,7 +316,9 @@ fn get_nodes(cluster_info: &ClusterInfo, stakes: &HashMap<Pubkey, u64>) -> Vec<N
     // The local node itself.
     std::iter::once({
         let stake = stakes.get(&self_pubkey).copied().unwrap_or_default();
-        let node = NodeId::from(cluster_info.my_contact_info());
+        let node = LegacyContactInfo::try_from(&cluster_info.my_contact_info())
+            .map(NodeId::from)
+            .unwrap();
```

Comment:
this one

Reply:
> This is clearly my_contact_info and can only fail if I set up my node with incorrect udp/quic offsets

this isn't my contention.  my contention is that the `my_contact_info()` cases use undocumented special knowledge about what is being returned to do something that might be dangerous on any other `ContactInfo`.  should a novice begin working in this area and use one of these instances as an example Best Practices :tm:, they may well introduce _unsafe_ `unwrap()`s.  now it is on the reviewer to a) know the undocumented special knowledge and b) discern whether the new code is safe or not.  this is a defensive coding request, not a this code is wrong request.  i understand that it is correct in this case, it's just setting an unsafe example for future devs. if we tie the special knowledge up in its own type, it's much more difficult to misuse

> should always succeed until the migration is fully done and LegacyContactInfo is removed from gossip

then we can switch to `From` ?

Reply:
We can't remove the `unwrap` here precisely because of the reason you are mentioning: "defensive coding".
If myself or someone else introduces a bug that breaks this, I want to stop them right there, not let them merge their code and not even let them get a green CI.

There are 2 different things:
1. One is the case that the code might panic because it is relying on some implementation details or other presumptions. We should identify and remove all such instances (just as I did in [#29805 deprecating `Pubkey::new`](https://github.com/solana-labs/solana/pull/29805)) and not introduce new ones.
2. The other one is that you are _enforcing some invariant in the code_ to stop bugs to be added in the future. There are **no** assumptions here. If the invariant ever fails to hold it means that someone has introduced a bug which should be reverted. We don't want to identify such bugs on mainnet. What the `unwrap` does is to make it almost impossible to even get a green CI.

The point is distinguishing (1) vs (2). We want the code to be so airtight that a bad change does not even pass the CI to be allowed to be merged in. This means (2) is a good thing, but it does not mean (1) is also then justified by any means; and the distinction can only be made on a case-by-case basis based on the very specifics of a particular usage.



>> should always succeed until the migration is fully done and LegacyContactInfo is removed from gossip
>
> then we can switch to From ?

no, a `ContactInfo` may or may not be converted to a `LegacyContactInfo`.
But I as a node operator should not spin up my node with any such `ContactInfo` which does convert to a `LegacyContactInfo`. The `unwrap` here catches and stops this latter case.

Reply:
> We can't remove the `unwrap` here precisely because of the reason you are mentioning: "defensive coding". If myself or someone else introduces a bug that breaks this, I want to stop them right there, not let them merge their code and not even let them get a green CI.

again, i'm not asking for the `unwrap()`s to be removed. i'm asking them to be [hidden behind a type](https://github.com/solana-labs/solana/pull/30022#pullrequestreview-1281697505) for the safe case of my contact info.

> There are 2 different things:
> 
> 1. One is the case that the code might panic because it is relying on some implementation details or other presumptions. We should identify and remove all such instances (just as I did in [#29805 deprecating `Pubkey::new`](https://github.com/solana-labs/solana/pull/29805)) and not introduce new ones.

is this not precisely the case being used to justify `unwrap()`/`expect()` on `my_contact_info()`.  it is assumed that the local node's `ContactInfo` was constructed correctly, or that we've already crashed at start up.

> 2. The other one is that you are _enforcing some invariant in the code_ to stop bugs to be added in the future. There are **no** assumptions here. If the invariant ever fails to hold it means that someone has introduced a bug which should be reverted. We don't want to identify such bugs on mainnet. What the `unwrap` does is to make it almost impossible to even get a green CI.

an `unwrap()` can only enforce an invariant if full test coverage is attained _and retained_, otherwise it's a target for some security researcher to get a DoS/LoA bounty out of us. given that the coverage ci step is basically only there to turn green, i'm loath to think this is something our dev processes are mature enough to assure, regardless of the confidence of any contributor at merge time.

> The point is distinguishing (1) vs (2). We want the code to be so airtight that a bad change does not even pass the CI to be allowed to be merged in. This means (2) is a good thing, but it does not mean (1) is also then justified by any means; and the distinction can only be made on a case-by-case basis based on the very specifics of a particular usage.
> 

agreed, but this is what _tests_ and type-safety are for, not sprinkling bullseyes all over the codebase for security researchers to aim at

> > > should always succeed until the migration is fully done and LegacyContactInfo is removed from gossip
> > 
> > 
> > then we can switch to From ?
> 
> no, a `ContactInfo` may or may not be converted to a `LegacyContactInfo`. But I as a node operator should not spin up my node with any such `ContactInfo` which does not convert to a `LegacyContactInfo`. The `unwrap` here catches and stops this latter case.

so which is it "should always succeed" or "may or may not be converted"?  `unwrap()` is not the tool for the job when it comes to validating user input (nor `expect()` or any other wrapper around `abort()`)


fwiw, i'm taking this hardline approach now, because i recognize that i'm going to need to take it when reviewing external contributions. i don't want to be playing favorites in the new world

Reply:
> again, i'm not asking for the `unwrap()`s to be removed. i'm asking them to be [hidden behind a type](https://github.com/solana-labs/solana/pull/30022#pullrequestreview-1281697505) for the safe case of my contact info.

Ok, if the `unwrap`s don't need to be removed then it shouldn't be a blocker to merge this change.
Instead of making this change bigger, we can discuss/evaluate/commit a change to improve the interface separately.
What you are suggesting [here](https://github.com/solana-labs/solana/pull/30022#pullrequestreview-1281697505) is not reducing number of unwraps but only hiding them; so that merits its own _separate_ conversation if that is a net positive change or we better keep things visible and explicit at call-sites.

> is this not precisely the case being used to justify `unwrap()`/`expect()` on `my_contact_info()`. it is assumed that the local node's `ContactInfo` was constructed correctly, or that we've already crashed at start up.

no, it is not making any such assumptions.

> an `unwrap()` can only enforce an invariant if full test coverage is attained _and retained_, otherwise it's a target for some security researcher to get a DoS/LoA bounty out of us. given that the coverage ci step is basically only there to turn green, i'm loath to think this is something our dev processes are mature enough to assure, regardless of the confidence of any contributor at merge time.

Removing such `unwrap`s which serve as sanity checks and safe guards would make it easier for bugs to creep in and resulting in more security bounties to be paid out at the end.
There is ample test coverage here and we can discuss how to improve and retain test coverage in general, but if the goal at the end is stopping bugs then removing such guardrails will not help with that.

> so which is it "should always succeed" or "may or may not be converted"? `unwrap()` is not the tool for the job when it comes to validating user input (nor `expect()` or any other wrapper around `abort()`)

* I as a node operator should not spin up my node with a `ContactInfo` which does not convert to a `LegacyContactInfo`.
* A node gossiping their `ContactInfo` to other nodes can send a `ContactInfo` which is not a valid `LegacyContactInfo`.

The distinction is if I am accessing _my_ contact-info or another node's contact-info.

Reply:
i believe we've reached an impasse

Reply:
> > again, i'm not asking for the `unwrap()`s to be removed. i'm asking them to be [hidden behind a type](https://github.com/solana-labs/solana/pull/30022#pullrequestreview-1281697505) for the safe case of my contact info.
> 
> Ok, if the `unwrap`s don't need to be removed then it shouldn't be a blocker to merge this change. Instead of making this change bigger, we can discuss/evaluate/commit a change to improve the interface separately. What you are suggesting [here](https://github.com/solana-labs/solana/pull/30022#pullrequestreview-1281697505) is not reducing number of unwraps but only hiding them; so that merits its own _separate_ conversation if that is a net positive change or we better keep things visible and explicit at call-sites.

What is the problem with above? :point_up: